Mathlib/Data/Real/StandardPart.lean

```diff
@@ -0,0 +1,320 @@
+/-
+Copyright (c) 2025 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+module
+
+public import Mathlib.Algebra.Order.Ring.Archimedean
+public import Mathlib.Algebra.Ring.Subring.Order
+public import Mathlib.Data.Real.Archimedean
+public import Mathlib.Data.Real.CompleteField
+public import Mathlib.Order.Quotient
+public import Mathlib.RingTheory.Valuation.ValuationSubring
+
+/-!
+# Standard part function
+
+Given a finite element in a non-archimedean field, the standard part function rounds it to the
+unique closest real number. That is, it chops off any infinitesimals.
+
+Let `K` be a linearly ordered field. The subset of finite elements (i.e. those bounded by a natural
+number) is a `ValuationSubring`, which means we can construct its residue field
+`FiniteResidueField`, roughly corresponding to the finite elements quotiented by infinitesimals.
+This field inherits a `LinearOrder` instance, which makes it into an Archimedean linearly ordered
+field, meaning we can uniquely embed it in the reals.
+
+Given a finite element of the field, the `ArchimedeanClass.standardPart` function returns the real
+number corresponding to this unique embedding. This function generalizes, among other things, the
+standard part function on `Hyperreal`.
+
+## Todo
+
+Redefine `Hyperreal.st` in terms of `ArchimedeanClass.standardPart`.
+
+## References
+
+* https://en.wikipedia.org/wiki/Standard_part_function
+-/
+
+@[expose] public section
+
+namespace ArchimedeanClass
+variable
+  {K : Type*} [LinearOrder K] [Field K] [IsOrderedRing K] {x y : K}
+  {R : Type*} [LinearOrder R] [CommRing R] [IsOrderedRing R] [Archimedean R]
+
+/-! ### Finite residue field -/
+
+variable (K) in
+/-- The valuation subring of elements in non-negative Archimedean classes, i.e. elements bounded by
+some natural number. -/
+protected noncomputable def Finite : ValuationSubring K :=
+  (addValuation K).toValuation.valuationSubring
+
+namespace Finite
+
+@[simp] theorem mem_finite_iff {x : K} : x ∈ ArchimedeanClass.Finite K ↔ 0 ≤ mk x := .rfl
+
+theorem mem_map_of_archimedean (f : R →+*o K) (r : R) : f r ∈ ArchimedeanClass.Finite K := by
+  obtain rfl | hr := eq_or_ne r 0
+  · simp
+  · rw [mem_finite_iff, mk_map_of_archimedean' f hr]
+
+@[simp]
+theorem mk_zero (h : 0 ∈ ArchimedeanClass.Finite K) :
+    (⟨0, h⟩ : ArchimedeanClass.Finite K) = 0 := rfl
+
+@[simp]
+theorem mk_one (h : 1 ∈ ArchimedeanClass.Finite K) :
+    (⟨1, h⟩ : ArchimedeanClass.Finite K) = 1 := rfl
+
+@[simp]
+theorem mk_neg (x : K) (h : -x ∈ ArchimedeanClass.Finite K) :
+    (⟨-x, h⟩ : ArchimedeanClass.Finite K) = -⟨x, neg_mem_iff.1 h⟩ := rfl
+
+@[simp]
+theorem mk_natCast (n : ℕ) (h : (n : K) ∈ ArchimedeanClass.Finite K) :
+    (⟨n, h⟩ : ArchimedeanClass.Finite K) = n := rfl
+
+@[simp]
+theorem mk_intCast (n : ℤ) (h : (n : K) ∈ ArchimedeanClass.Finite K) :
+    (⟨n, h⟩ : ArchimedeanClass.Finite K) = n := rfl
```

Comment:
These lemmas are really weird. These are general lemmas about `ValuationSubring`s, and I don't expect we would need them.